pvanalytics/features/clipping.py

```diff
@@ -88,3 +88,141 @@ def levels(ac_power, window=4, fraction_in_window=0.75,
+
+    return daytime_power.groupby(
+        daytime_power.index.hour * 60 + daytime_power.index.minute
+    ).quantile(power_quantile)
+
+
+def _clipped(power, derivative, power_min, derivative_max):
+    # return a mask that is true where `power` is greater than
+    # `power_min` and the absolute value of `derivative` is less
+    # than `derivative_max`
+    return (np.abs(derivative) <= derivative_max) & (power > power_min)
+
+
+def _clipping_power(ac_power, derivative_max=0.0035, power_min=0.75,
+                    power_quantile=0.995, frequency_quantile=0.25,
+                    freq=None):
+    # Calculate a power threshold, above which the power is being
+    # clipped.
+    #
+    # - The daytime power curve is calculated using
+    #   _daytime_powercurve(). This function groups `ac_power` by
+    #   minute of the day and returns the `power_quantile`-quantile of
+    #   power for each minute, excluding night time, early morning,
+    #   and late afternoon/evening (the amount of data that is
+    #   excluded is controlled by `frequency_quantile`).
+    #
+    # - Each timestamp in the daytime power curve that satisfies the
+    #   clipping criteria[*] is flagged.
+    #
+    # - The clipping threshold is calculated as the mean of the
+    #   longest flagged period in the daytime power curve
+    #
+    # [*] clipping criteria: a timestamp satisfies the clipping
+    # criteria if the derivative of the daytime power curve at that
+    # time is less than `derivative_max` and the value of the daytime
+    # power curve is greater than `power_min` times the median of the
+    # daytime power curve.
+    #
+    # Based on the PVFleets QA Analysis project
+    if not freq:
+        freq = pd.Timedelta(pd.infer_freq(ac_power.index)).seconds / 60
+    elif isinstance(freq, str):
+        freq = pd.Timedelta(freq).seconds / 60
+
+    # Use the derivative of the 99.5% quantile of daytime power at
+    # each minute to identify clipping.
+    powercurve = _daytime_powercurve(ac_power)
+    normalized_power = powercurve / powercurve.max()
+    power_derivative = (normalized_power.diff()
+                        / normalized_power.index.to_series().diff()) * freq
+
+    clipped_times = _clipped(powercurve, power_derivative,
+                             powercurve.median() * power_min,
+                             derivative_max)
+    clipping_cumsum = (~clipped_times).cumsum()
+    # get the value of the cumulative sum the longest True span
+    longest_clipped = clipping_cumsum.value_counts().idxmax()
+    # select the longest span that satisfies the clipping criteria
+    longest = powercurve[clipping_cumsum == longest_clipped]
+
+    if longest.index[-1] - longest.index[0] >= 60:
+        # if the period of clipping is at least 60 minutes then we
+        # have enough evidence to determine the clipping threshold.
+        return longest.mean()
+    return None
+
+
+def threshold(ac_power, derivative_max=0.0035,
```

Comment:
I'm inclined to ask for `derivative` and `derivative_max` to be `slope` and `slope_max` respectively. Slope is (to me) unambiguous, whereas a numerical derivative can become quite complicated in its calculation.

Reply:
Sounds good to me.